8.0: add purchase_requisition_bid_selection

lepistone opened the pull request:
See `__openerp__.py` for details.

Thanks!

jbaudoux committed:
[ADD] purchase and purchase requisition generic improvements
(../generic-addons/ rev 10.1.1)

Commit:
[FIX] wrong company name, missing license, missing coding, wrong indentation for brackets (pep-8)
(../generic-addons/ rev 10.2.1)

Commit:
[FIX] wrong company name, missing license, missing coding line
(../generic-addons/ rev 10.2.7)

Commit:
[FIX] orm instead of osv, orm.Model instead of osv.osv and except_orm instead of except_osv (deprecated)
(../generic-addons/ rev 10.2.8)

Commit:
[FIX] pep8: space after commas, spaces around operators, cut long line, too short variable name; removed dead code
(../generic-addons/ rev 10.2.9)

Commit:
[CHG] set item instead of update() when updating only 1 item, much more efficient
(../generic-addons/ rev 10.2.10)

Commit:
[FIX] intricated condition
(../generic-addons/ rev 10.2.11)

Commit:
[FIX] early return instead of nesting all the method
(../generic-addons/ rev 10.2.12)

Commit:
[FIX] public method are not allowed to return None
(../generic-addons/ rev 10.2.13)

Commit:
[FIX] assert the requirement of only 1 id when the ids can be a list or tuple instead of just taking the first one, which can be a misleading behavior
(../generic-addons/ rev 10.2.14)

Commit:
[FIX] typo: 'canceled' with 1 'l' tends to be the used form in US
(../generic-addons/ rev 10.2.15)

Commit:
[FIX] simplify really intricated method, it was hard to read and to understand what it was doing
(../generic-addons/ rev 10.2.16)

Commit:
[FIX] typo: openerp.osv not openerp.orm of course
(../generic-addons/ rev 10.2.17)

jgrandguillaume committed:
[REVIEW] reviewed feature branch: pep8, styling, rewrite of intricate methods, licenses, ...
(../generic-addons/ rev 10.1.3)

jbaudoux committed:
[IMP] moved purchase modifications to the right module
      [IMP] improved bid workflow
(../generic-addons/ rev 15.5.1)

jbaudoux committed:
[FIX]purchase_extended: use rfq/po type in menu actions to differentiate cancelled rfq/po
(../generic-addons/ rev 15.5.3)

jbaudoux committed:
[IMP]purchase_requisition_extended: add button to view generated po and fix button to view rfq/bids
(../generic-addons/ rev 15.5.4)

jbaudoux committed:
[IMP]purchase_requisition: At PO generation, cancell only rfq not the bids
(../generic-addons/ rev 15.5.5)

jbaudoux committed:
[IMP]purchase_requisition: rename button Done into Generate PO
(../generic-addons/ rev 15.5.6)

jbaudoux committed:
[FIX] after rebase and code review
(../generic-addons/ rev 15.5.7)

jbaudoux committed:
[IMP] purchase_requisition: Bids comparison: show only bids encoded and group by call for bids line
(../generic-addons/ rev 15.5.8)

jbaudoux committed:
[IMP]purchase_extended, purchase_requisition: documentation
(../generic-addons/ rev 15.5.9)

jbaudoux committed:
[ADD]purchase: reason for cancellation + few fixes
(../generic-addons/ rev 15.5.10)

jbaudoux committed:
[IMP]purchase, purchase_requisition: add state draft bid for registering bids in case of open call for bids + add check call for bids has lines when confirming
(../generic-addons/ rev 15.5.13)

jbaudoux committed:
[FIX]purchase: renamed cancelled into canceled
(../generic-addons/ rev 15.5.14)

jbaudoux committed:
[FIX]purchase: replaced osv.except_osv by orm.except_orm, osv.osv by osv.Model, osv.osv_memory by osv.TransientModel
(../generic-addons/ rev 15.5.16)

jbaudoux committed:
[FIX] pep8
(../generic-addons/ rev 15.5.17)

jbaudoux committed:
[FIX]purchase, purchase_requisition: add state draft bid for registering bids in case of open call for bids: forget to add some files
(../generic-addons/ rev 15.5.18)

jbaudoux committed:
[FIX]purchase_requisition: fix PO line name_get if no scheduled date
(../generic-addons/ rev 15.5.19)

jbaudoux committed:
[ADD] YML tests + few fixes
(../generic-addons/ rev 15.5.20)

jbaudoux committed:
[FIX]purchase_requisition: fix confirm selection
(../generic-addons/ rev 15.5.22)

jbaudoux committed:
[MOD]purchase_requisition: bid tendering mode (open/restricted): removed default value and set as required in view only
(../generic-addons/ rev 15.5.23)

jbaudoux committed:
[FIX]purchase: on_change product must manage price properly with updated workflow
(../generic-addons/ rev 15.5.26)

jbaudoux committed:
[FIX]purchase_requisition: fixed float comparision rounding issue
(../generic-addons/ rev 15.5.27)

jbaudoux committed:
[FIX]purchase: noupdate for cancel reason data
(../generic-addons/ rev 15.5.28)

jbaudoux committed:
[FIX]purchase: context shouldn't be modified directly without doing a copy
(../generic-addons/ rev 15.5.29)

nbessi committed:
[MRG] Major modification of procurement workflow for purchase and sale. Please see nested history for details
(../generic-addons/ rev 33)

jbaudoux committed:
[IMP]purchase_requisition: do not aggregate price and qty in bid comparison
(../generic-addons/ rev 15.5.31)

jbaudoux committed:
[FIX]import, orm
(../generic-addons/ rev 15.5.32)

jbaudoux committed:
[IMP]purchase_requisition: do not aggregate price and qty in bid comparison
  [FIX]purchase: invoice button should not be visible on draft rfq/bid/po
(../generic-addons/ rev 37)

jbaudoux committed:
[FIX]purchase_requisition: fix cancel reason
(../generic-addons/ rev 15.5.33)

jbaudoux committed:
[FIX]purchase_requisition: remove sum on lead time in Bid Analysis
(../generic-addons/ rev 15.5.34)

Commit:
[MRG] from master branch
(../generic-addons/ rev 29.1.1)

jbaudoux committed:
[FIX]purchase_requisition: remove sum on lead time in Bid Analysis, fix cancel reason
(../generic-addons/ rev 38)

Commit:
[CHG] blocks selection of more qty of purchase lines than requested or no lines selected, warn if less
(../generic-addons/ rev 29.1.4)

Commit:
[FIX] move methods which have nothing to do in purchase.order.line but are related to purchase.requisition
(../generic-addons/ rev 29.1.5)

Commit:
[MRG] from master branch
(../generic-addons/ rev 29.1.6)

Commit:
[FIX] give details on why we check the 'ids' argument (due to an error of the webclient which do not send ids in some cases)
(../generic-addons/ rev 29.1.7)

Commit:
[CHG] allow to select more lines than requested in bids, display a warning instead of blocking the user
(../generic-addons/ rev 29.1.12)

jbaudoux committed:
[ADD]purchase_requisition: yml test file of a complete restricted call for bids process
(../generic-addons/ rev 29.2.1)

Commit:
[CHG] allow again to select no quantity for a line of a purchase requisition
(../generic-addons/ rev 29.1.15)

Commit:
[IMP] simplify the warning displayed when the selection of lines does not match the requested quantity
(../generic-addons/ rev 29.1.16)

jgrandguillaume committed:
[MRG] [IMP] Extract methods for re-usability, fixes
  [ADD] sale_exception_nostock_by_line
  Summary of what I have done here:
  - Describe the sale_ownership module
  - extract methods related to purchase_requisition from purchase_order_line
   - sale_dropshipping: extracted method to allow re-usability (& took opportunity for cleanings and fixes of views)
   - display a different message when we choose less / greater / no quantity in purchase requisitions
(../generic-addons/ rev 39)

jgrandguillaume committed:
[MRG] [ADD]purchase_requisition: yml test file testing a complete restricted call for bids process from the RFQ creation and bids encoding until the comparison and bids selection and generation of the PO.
(../generic-addons/ rev 40)

jgrandguillaume committed:
[DEL] Uneeded files in purchase_requisition that were still here after replay the revision with the community script

nbessi committed:
[FIX] PO cancelation when cost estimate is confirmed

Commit:
[IMP] add Bid selected state on PO

jgrandguillaume committed:
[ADD] The pricelist on the requisition and bid creation process

rdeheele committed:
[FIX] overwrite copy to pick origin if po comes from requisition

rdeheele committed:
[IMP] improve origin checking

jgrandguillaume committed:
[IMP] Add a pricelist in tendering process

rdeheele committed:
[IMP] add tooltips on 4 fields

rdeheele committed:
[FIX] forget a condition : only set origin if the PO comes from requisition

yvaucher committed:
[IMP] assign single dict value using key instead of using update

nbessi committed:
[FIX] merge trouble

nbessi committed:
[FIX] typo

nbessi committed:
[FIX] purchase price list only should be available in selection

nbessi committed:
[FIX] cancel of requisition did not cancel related PO

nbessi committed:
[REFACTOR] purchase.requisition cancelation in order to be code polite and overridable

nbessi committed:
[FIX] unused parren

yvaucher committed:
add license header

yvaucher committed:
port models and methods to api 8.0

yvaucher committed:
fix _get_tender

gurneyalex committed:
proof read module description

gurneyalex committed:
fix wrong import

gurneyalex committed:
remove spurious comma which created tuple

... and broke the column definition

gurneyalex committed:
reformat

lepistone committed:
WIP: make module installable

for the moment, disable broken things from the views

lepistone committed:
fix renamed class purchase.cancel_reason

lepistone committed:
make yml test work (but not pass) on v8

lepistone committed:
port wizard

lepistone committed:
WIP port purchase_requisition_extended

lepistone committed:
improve a failing assertion

lepistone committed:
lint manifest

lepistone committed:
remove duplicate field

lepistone committed:
get rid if _get_tender and make it related

- some teething issues with the new API
- this code looks very much like a simple related field, so I am not
  porting that to the new API since maybe we could get rid of the code.

lepistone committed:
show req_validity on view

lepistone committed:
use asustek instead of agrolait: it is a supplier

lepistone committed:
fix button to show product lines

lepistone committed:
improve yml test

lepistone committed:
improve view

lepistone committed:
shorten xml ids in tests

lepistone committed:
refactor test to help me figure it out

lepistone committed:
make purchase_ids classic again

lepistone committed:
adapt to renamed purchase_extended

lepistone committed:
rename purchase_requisition_bid_extended

was purchase_requisition_extended

lepistone committed:
fix type and state of generated PO

lepistone committed:
update copyright

lepistone committed:
lint

lepistone committed:
fix view, onchange

lepistone committed:
fix onchange picking type

yvaucher committed:
port get_object_reference to xmlid_to_object as it raise a ValueError as it was half ported by removing cr uid

yvaucher committed:
fix generate_po check to ensure it is executed on only one rec

jgrandguillaume commented:
Hi,

I've tested the module with the following test case:

 1) Create a tender for 1000 Apple keyboard and confirm
 2) Create 2 RFQ using the buton "encode bid" for Camptocamp and Axelor
 3) The RFQ are created to the state "Draft bid" and the next action is bid encoded. The state should normally be "draft RFQ".

At this stage, I should have 2 RFQ with the state "Draft RFQ" and the next action should be to send the RFQ by mail. The state should then pass to RFQ sent. Then the supplier send their bids and we should encode them, pushing the "Bid encoded" button and the state should then be "bid encoded" (so we should never have the "draft bid stage").
This is the standard process of RFQ introduced by "purchase_rfq_bid_workflow" module if you don't pass through a tender:
- draf RFQ
- RFQ sent
- bid encoded
- draft PO

Then, If I continue the process anyway:

 4) I encode the bids by putting a price on lines and clicking "Bid encoded". A popup ask me to give the reception date  and the state pass to "bid encoded".

=> Here in the popup, I see 2 "Cancel" button and I should have only one

 5) Then, once my 2 bids are in state "Bid encoded" I close the requisition by clicking "Close call for bid" the state of the requisition pass to "Bid selection"

=> Here the button "Close call for bid" should be in red

 6) Then I should chose my bid by clicking "Chose product lines". It open a view where I can select lines. Here I select the whole quantity from the first line by clicking the green arrow. I then click the "Generate PO" Button (top left).

=> Here it get me back to the requisition view, but no PO has been created... and it should have. The state of the requisition should also have moved from "Bids selection" to "PO Created". The only button left here is "Chose product Line" and I'm stuck here...

I suggest here the following for 6):

a) Click "Chose product line" It open a view where I can select lines. Here I select the whole quantity from the first line by clicking the green arrow. I then click a new button "Bis selected" that bring me back to the requisition and move his stage to "Bids selected"

b) I have a button on the requisition "Generate PO" that create the PO from the chosen bids. The bids should remain at sate "Bid encoded" and new PO in draft state should be generated. The state of the requisition should then be "PO created"

I can't remember exactly, but I know last year we had the generate PO only that does the job a close the requisition by generating the PO immediatly, skipping the bid selected state. But I think, it is better to have all state.

Hope this is clear enough.

Thanks for the good job.

Regards,

Joël

lepistone commented:
Thanks for testing @jgrandguillaume. Are you testing on open or restricted mode? I think in open mode it is correct to enter directly a bid.

I will continue dealing with this PR tomorrow.

jgrandguillaume commented:
Hi @lepistone , 

You're perfectly right on that point. In "Open" mode, the bid arrives directly encoded. In restricted mode, you encode the RFQ. So my point 3) above isn't correct.

Regards,

Joël

lepistone commented:
Hi @jgrandguillaume.
- 1-3 It's already OK
- 4: I cannot reproduce the double cancel button, can you give more info?
- 5 The button is red in fact, but becomes grey in edit mode. That is a strange behaviour indeed, but the core purchase_requisition does the same. I think it is unrelated to that module
- 6 PO creation is covered by a yaml test that passes. However you are right in saying that the interface for that is not working correctly. That view is done in javascript in the core, and again modified in javascript in that module, and I need some help to deal with it. Can you suggest someone to help me out?

lepistone committed:
fix copy if type='purchase'

If a purchase.order is copied with type=purchase, make copy() calls
create() with 'draft_po' in the context. This way create() takes care of
advancing the workflow appropriately to draft_po.

jgrandguillaume commented:
Hi @lepistone ,

Thanks for the feed-back ! Let's say 1-5 are ok then. I suggest we focus on 6 as it is the most problematic point (cannot ends the workflow currently).

@gurneyalex will have a look with you here (I already checked with him)

Regards,

Joël

gurneyalex committed:
fix js file inclusion

in v8, js/css files are no longer included from __openerp__.py, but by updating
the qweb template

gurneyalex committed:
fix call to close_callforbids

the id of the purchase requisition is no longer in active_ids but in tender_id,
passed in the context.

gurneyalex committed:
add missing check for default is None

jgrandguillaume commented:
Hi,

Thanks for the fix ! It's working fine now ! 

:+1: 

Regards,

Joël

gurneyalex committed:
fix flake8 warning

gurneyalex commented:
fixed remaining issues from @jgrandguillaume,  fixed tests, fixed flake8

:+1:

lepistone commented:
Thanks for your help @jgrandguillaume @gurneyalex! It looks like we're good now.

yvaucher commented:
:+1:
